### PR TITLE
Cleanup Session/Authentication

### PR DESCRIPTION
**Pull request by nupplaphil**
`Authentication::withSession()` - for auto authentication with Session/Cookie variables
`Authentication::withOpenId()` - for authentication with an OpenID account
`Authentication::withPassword()` - for authentication with Password
`Authentication::setForUser()` - for setting the user auth context of the current session

Refactor `Session` class - contains now only "native" Session Management methods

=> Tested local with PW/logout/login/site-navigation
=> Tested on my node with 2Factor on

This is a prerequisite for further changes around the `Session` class

### CODE REVIEW

**Comment by nupplaphil** [comment]
Needs https://github.com/friendica/friendica-addons/pull/917 as well

File: src/Core/Authentication.php [hunk 1]
```diff
@@ -6,69 +6,378 @@
...
+use Friendica\Util\Strings;
+use LightOpenID;
+use Friendica\Core\L10n\L10n;
+use Psr\Log\LoggerInterface;
 
 /**
  * Handle Authentification, Session and Cookies
  */
-class Authentication extends BaseObject
+class Authentication
```
Comment (MrPetovan) [inline src/Core/Authentication.php]: Almost all of the method of this new class take and `App` as first argument, which makes me think these methods should be moved to `App`.

File: src/Core/Session.php [hunk 2]
```diff
@@ -278,4 +170,75 @@ public static function isAuthenticated()
...
+
+	/**
+	 * @brief Calculate the hash that is needed for the "Friendica" cookie
+	 *
+	 * @param array $user Record from "user" table
+	 *
+	 * @return string Hashed data
+	 * @throws \Friendica\Network\HTTPException\InternalServerErrorException
+	 */
+	private static function getCookieHashForUser($user)
```
Comment (MrPetovan) [inline src/Core/Session.php]: Shouldn't this be moved to `Model\User`?

File: src/Core/Session.php [hunk 3]
```diff
@@ -104,117 +102,11 @@ public static function remove($name)
 	 */
 	public static function clear()
 	{
+		session_unset();
```
Comment (MrPetovan) [inline src/Core/Session.php]: What is the point of adding these two lines?

**Comment by annando** [comment]
We are short before the RC. In this phase we only should add some guaranteed non blocking PRs. This here seems to me like something that really needs testing. Having it in the next release (not this) would be better, I guess.

**Comment by MrPetovan** [comment]
Agreed.

**Comment by nupplaphil** [comment]
Yes, you're right. It was in my backlog for a long time and finally I found some minutes yesterday to finish it ;-) .. So I change the milestone and wait for the RC branch before merging it in the dev?

File: src/Core/Session.php (hunk 3 above)
Comment (nupplaphil) [reply src/Core/Session.php]: I'm not sure about backwards compatibility in combination with `session_start()`, so I decided to use some older functions as well. Do you think this isn't necessary?

File: src/Core/Session.php (hunk 2 above)
Comment (nupplaphil) [reply src/Core/Session.php]: I moved it to a own class `Model\User\Cookie` so we don't create "monster" classes :-)

File: src/App/Authentication.php [hunk 4]
```diff
@@ -44,63 +50,62 @@ class Authentication
...
-					]
-				);
-				if (DBA::isResult($user)) {
-					if (!Session::checkCookie($data->hash, $user)) {
-						$this->logger->notice("Hash doesn't fit.", ['user' => $data->uid]);
-						Session::delete();
-						$this->baseUrl->redirect();
-					}
+		// When the "Friendica" cookie is set, take the value to authenticate and renew the cookie.
+		if (isset($data) && isset($data->uid)) {
```
Comment (MrPetovan) [inline src/App/Authentication.php]: This can be shortened as `if (isset($data->uid)) {` and mean the same.

File: src/Model/User.php [hunk 5]
```diff
@@ -9,12 +9,14 @@
 
 use DivineOmega\PasswordExposed;
 use Exception;
+use Friendica\App;
```
Comment (MrPetovan) [inline src/Model/User.php]: ?

File: src/Model/User.php [hunk 6]
```diff
@@ -9,12 +9,14 @@
...
 use Friendica\Core\Config;
 use Friendica\Core\Hook;
 use Friendica\Core\L10n;
 use Friendica\Core\Logger;
 use Friendica\Core\PConfig;
 use Friendica\Core\Protocol;
+use Friendica\Core\Session;
```
Comment (MrPetovan) [inline src/Model/User.php]: ??

File: src/Model/User/Cookie.php [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+	 * @param string $value    [optional]
+	 * @param int    $expire   [optional]
+	 * @param string $path     [optional]
+	 * @param string $domain   [optional]
+	 * @param bool   $secure   [optional]
+	 * @param bool   $httponly [optional] <p>
+	 *
+	 * @return bool If output exists prior to calling this function,
+	 *
+	 * @since 4.0
```
Comment (MrPetovan) [inline src/Model/User/Cookie.php]: What are these version numbers about?

File: src/Core/Session.php (hunk 3 above)
Comment (MrPetovan) [reply src/Core/Session.php]: There seem to be a `destroy` method that does the `session_unset` (but not the `session_start`) so I was wondering if you have a specific reason for adding these two lines here. I don't think it is necessary,

**Comment by MrPetovan** [comment]
Waiting RC branch fork to merge into `develop`.

**Comment by nupplaphil** [comment]
So, now I'm finished .. Originally, I wanted to first merge the Auth refactoring and then the Session Management, but now I added both :)

After the merge, we can now better control the Session behavior for backend processes and testing.

**Comment by nupplaphil** [comment]
And now we don't need the App class for a correct session initializing anymore ;)

File: src/Factory/SessionFactory.php [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+namespace Friendica\Factory;
+
+use Friendica\App;
+use Friendica\Core\Cache\Cache;
+use Friendica\Core\Cache\ICache;
+use Friendica\Core\Config\Configuration;
+use Friendica\Core\Session\CacheSession;
+use Friendica\Core\Session\DatabaseSession;
+use Friendica\Core\Session\ISession;
+use Friendica\Core\Session\Memory;
```
Comment (MrPetovan) [inline src/Factory/SessionFactory.php]: Superfluous `use` statement?

File: src/Factory/SessionFactory.php [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+use Friendica\Core\Session\MemorySession;
```
Comment (MrPetovan) [inline src/Factory/SessionFactory.php]: Now that you renamed the classes, this looks wrong.

File: src/Factory/SessionFactory.php [hunk 10]
```diff
@@ -6,12 +6,12 @@
...
 use Friendica\Core\Config\Configuration;
-use Friendica\Core\Session\CacheSession;
-use Friendica\Core\Session\DatabaseSession;
+use Friendica\Core\Session\Cache;
+use Friendica\Core\Session\Database;
 use Friendica\Core\Session\ISession;
 use Friendica\Core\Session\Memory;
-use Friendica\Core\Session\MemorySession;
-use Friendica\Core\Session\NativeSession;
+use Friendica\Core\Session\Memory;
```
Comment (MrPetovan) [inline src/Factory/SessionFactory.php]: Duplicate use statement.

File: src/Core/Session/Memory.php [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+	/**
+	 * @inheritDoc
+	 */
+	public function delete()
+	{
+		$this->data = [];
+		return true;
+	}
+}
```
Comment (MrPetovan) [inline src/Core/Session/Memory.php]: Code standards: Missing EOL EOF

**Comment by MrPetovan** [comment]
Tests are failing.

**Comment by nupplaphil** [comment]
locally tests are working.. But travis is currently really instable .. I'll wait until I can see the travis test results ..

File: src/Core/Session/Memory.php [hunk 12]
```diff
@@ -2,13 +2,21 @@
...
+use Friendica\Model\User\Cookie;
+
 /**
  * Usable for backend processes (daemon/worker) and testing
  *
  * @todo after replacing the last direct $_SESSION call, use a internal array instead of the global variable
  */
 final class Memory extends Native
 {
+	public function __construct(Configuration $config, Cookie $cookie)
```
Comment (MrPetovan) [inline src/Core/Session/Memory.php]: You don't use `$config`?

**Comment by nupplaphil** [comment]
I reverted some of my changes. The DB and Cache handler are now handler again and no more own instances. I think this was more complicated than necessary.

**Comment by nupplaphil** [comment]
Ping @mrpetovan ^^

**Pull request merged**